designs/NuGetExtensibility/INuGetProjectServices.md

```diff
@@ -0,0 +1,138 @@
+
+See [NuGet extensibility services](NuGetExtensibilityServices.md) for more details.
+
+## Who are the customers
+
+Visual Studio component and extension developers.
+
+## Goals
+
+* Can be called on any thread (ideally background thread, but work if called on UI thread).
+* Do not block caller (implement async properly, return incomplete task instead).
+* Works with Visual Studio Codespaces, and hopefully Live Share.
+
+## Non-Goals
+
+* This will not be a 1:1 replacement for any of NuGet's existing `IVs` interfaces.
+
+## Solution
+
+```cs
+/// <summary>Service to interact with projects in a solution</summary>
+public interface INuGetProjectServices
+{
+    /// <Summary>Gets the list of packages installed in a project.</Summary>
+    /// <param name="projectId">Project ID (GUID).</param>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>
+    /// The list of packages in the project.
+    /// </returns>
+    /// <exception cref="System.ArgumentException">When projectId is not a guid.</exception>
+    Task<GetInstalledPackagesResult> GetInstalledPackagesAsync(string projectId, CancellationToken cancellationToken);
+}
+
+/// <summary>Result of a call to INuGetProjectServices.GetInstalledPackagesAsync</summary>
+public sealed class GetInstalledPackagesResult
+{
+    /// <summary>The status of the result</summary>
+    public GetInstalledPackageResultStatus Status { get; }
+
+    /// <summary>List of packages in the project</summary>
+    /// <remarks>May be null if <see cref="Status"/> was not successful</remarks>
+    public IReadOnlyCollection<NuGetInstalledPackage> Packages { get; }
+
+    // To maximize the chance this class can be converted to a C#9 record type without breaking changes, constructor is not public
+    internal GetInstalledPackagesResult();
+}
+
+/// <summary>The status of the result</summary>
+public enum GetInstalledPackageResultStatus
+{
+    /// <summary>Unknown status</summary>
+    /// <remarks>Probably represents a bug in the method that created the result.</remarks>
+    Unknown = 0,
+
+    /// <summary>Successful</summary>
+    Successful,
+
+    /// <summary>The project is not yet ready</summary>
+    /// <remarks>This typically happens shortly after the project is loaded, but the project system has not yet informed NuGet about package references</remarks>
+    ProjectNotReady,
+
+    /// <summary>Package information could not be retrieved because the project is in an invalid state</summary>
+    /// <remarks>If a project has an invalid target framework value, or a package reference has a version value, NuGet may be unable to generate basic project information, such as requested packages.</remarks>
+    ProjectInvalid
+}
+
+/// <summary>Basic information about a package</summary>
+public sealed class NuGetInstalledPackage
```

Comment:
Looks like we might need the PackagePath but you can verify that with the test team. https://github.com/NuGet/NuGet.Client/blob/f95e1aff6ceb901706ed19cd6cbd63ec05d40b07/src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPackageInstallerServices.cs#L93